apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -403,214 +433,118 @@ deliver(_ClientInfo, _Delivers, Session) ->
 handle_timeout(
     ClientInfo,
     ?TIMER_PULL,
-    Session0 = #{
-        id := Id,
-        inflight := Inflight0,
-        subscriptions := Subs,
-        props := Conf,
-        receive_maximum := ReceiveMaximum
-    }
+    Session0
 ) ->
-    MaxBatchSize = emqx_config:get([session_persistence, max_batch_size]),
-    BatchSize = min(ReceiveMaximum, MaxBatchSize),
-    UpgradeQoS = maps:get(upgrade_qos, Conf),
-    PreprocFun = make_preproc_fun(ClientInfo, Subs, UpgradeQoS),
-    {Publishes, Inflight} = emqx_persistent_message_ds_replayer:poll(
-        PreprocFun,
-        Id,
-        Inflight0,
-        BatchSize
-    ),
-    IdlePollInterval = emqx_config:get([session_persistence, idle_poll_interval]),
+    {Publishes, Session1} = drain_buffer(fetch_new_messages(Session0, ClientInfo)),
     Timeout =
         case Publishes of
             [] ->
-                IdlePollInterval;
+                emqx_config:get([session_persistence, idle_poll_interval]);
             [_ | _] ->
                 0
         end,
-    Session = emqx_session:ensure_timer(?TIMER_PULL, Timeout, Session0#{inflight := Inflight}),
+    Session = emqx_session:ensure_timer(?TIMER_PULL, Timeout, Session1),
     {ok, Publishes, Session};
-handle_timeout(_ClientInfo, ?TIMER_GET_STREAMS, Session) ->
-    renew_streams(Session),
+handle_timeout(_ClientInfo, ?TIMER_GET_STREAMS, Session0 = #{s := S0}) ->
+    S = emqx_persistent_session_ds_stream_scheduler:renew_streams(S0),
     Interval = emqx_config:get([session_persistence, renew_streams_interval]),
-    {ok, [], emqx_session:ensure_timer(?TIMER_GET_STREAMS, Interval, Session)};
-handle_timeout(_ClientInfo, ?TIMER_BUMP_LAST_ALIVE_AT, Session0) ->
+    Session = emqx_session:ensure_timer(
+        ?TIMER_GET_STREAMS,
+        Interval,
+        Session0#{s => S}
+    ),
+    {ok, [], Session};
+handle_timeout(_ClientInfo, ?TIMER_BUMP_LAST_ALIVE_AT, Session0 = #{s := S0}) ->
+    S = emqx_persistent_session_ds_state:commit(bump_last_alive(S0)),
```

Comment:
I'd expects users will want this configurable, especially when QoS2 flows are involved.

Reply:
Not sure if I understand the remark. `bump_interval()` function inside `bump_last_alive()` reads the configuration. `?TIMER_BUMP_LAST_ALIVE_AT` is the id of the timer event.

Reply:
Sorry, I was thinking about the configurability of the circumstances under which the session state is persisted.

Reply:
I specifically tried to avoid adding new config parameters in this PR, since I find constant adding and deprecation of config parameters quite annoying.
 
If there is real-life need to fine tune these parameters separately, we can always add a new timer/new config.